### DISCUSSION TITLE
Getting Intensity of a pixel

### DISCUSSION THREAD

**Discussion by hbckleikamp** [scholi/pySPM#35] [Q&A]
Dear scholi, thanks a lot for this nice module.
I am currently trying to get the intensity values from each mass-channel in each pixel.

I used this the code below, based on your module to get the counts of each channel per pixel.
But now I am wondering if I can also get the channel intensity for each pixel, is this also possible with pySPM, or am I misunderstanding the layout of .itm files?

Kind regards and many thanks, 
Hugo.

```
I = pySPM.ITM(ITM_file)
scans=np.arange(I.Nscan)
number_channels = int(round(A.root.goto('propend/Measurement.CycleTime').get_key_value()['float'] \
                            / A.root.goto('propend/Registration.TimeResolution').get_key_value()['float']))
    

for s in scans:
    raw = I.get_raw_raw_data(s)
    rawv = np.array(struct.unpack('<{}I'.format(len(raw) // 4), raw)).flatten()
        
    coords=np.argwhere(rawv>number_channels).flatten()
    data_start=coords[np.argwhere(np.diff(coords)>1).flatten()][1:]-2
    
    pixs=np.array_split(rawv,data_start)
    xs,ys,ms=[],[],[] #xcoords, ycoords, channels (masses)
    
    for c,p in enumerate(pixs): #loop over pixels
        xs.append(p[0] & 0x0fffffff)
        ys.append(p[1] & 0x0fffffff)        
        ms.append(p[p<number_channels]) #not sure if this is correct?
```

**Comment by scholi** [comment]
Hi Hugo,
What do you exactly means by the pixel intensity? The total counts for all masses? Did you try `I.get_intensity()`

Edit: by looking more into detail to your code, I'm not sure what you are trying to do. The `I.get_raw_raw_data()` gives you the very raw data which are not really useful to the end user. I believe what you are looking for is `I.get_raw_data()` which returns dictionary `d` where the key is the pixel corrdinates `(x,y)` and the value a list of the channels for the givel pixel. I have the imperssion you are trying to replicate this function somehow....

**Reply by hbckleikamp** [reply]
Dear Olivier, thank you for the response. Yes you are correct about me trying to replicate your I.get_raw_data() function. Sorry if my question is stupid, but I am new to these file types.

In the documentation of surfacelab about the new .itmx format the word "intensity" is used:
"The Extended Measurement *.itmx is a new format of SurfaceLab 7 for fast access to any
information in five dimensions (mass, intensity, x, y, z) (Interactive Data Analysis)."

I am trying to get .mgf-like output files for processing with metabolomics softwares.
Here, the first column of each scan represents mass and the 2nd column represents intensity.

Would such a file generated from an .itm file then look like this:

case 1 (with summed channel counts):
BEGIN IONS
PEPMASS=413.26611887841
CHARGE=1+
TITLE=scan n at pixel x,y and depth z
189.48956 1
283.62076 2
301.22977 1
311.08008 1
399.99106 1
END IONS

or this:

case 2 (with intensity):
BEGIN IONS
PEPMASS=413.26611887841
CHARGE=1+
TITLE=scan n at pixel x,y and depth z
189.48956 142145
283.62076 2661238
301.22977 5698
311.08008 4928
399.99106 1104
END IONS

Many thanks,
Hugo

**Reply by Said78-dot** [reply]
Hello how to get a pySPM module with spm_file composante
